### DISCUSSION TITLE
Having issues with the most recent release, output doesn't automatically calculate formulas

### DISCUSSION THREAD

**Discussion by allredbm** [PHPOffice/PhpSpreadsheet#2390] [Q&A]
The issue I'm having is twofold. I'm using PHP 8.0.3 and, while I don't know which version of PHPSpreadsheet I'm using, it's only a month or two out of date.

The first issue is basically exactly like the issue described here: https://github.com/PHPOffice/PhpSpreadsheet/issues/456. I have a basic PHP program that instantiates a reader, uses it to read an existing .xlsx file and create a Spreadsheet, instantiates a writer, and uses the Spreadsheet to output a copy of the original. The original spreadsheet has formulas that are designed to automatically calculate when certain other fields are filled, as per usual excel formula behavior. The problem is that the outputted copy does not automatically calculate these formulas like the original does. I saw that the issue above addressed the issue and pushed a solution into the codebase, but that solution doesn't seem to be helping here. I also tried the solution mentioned in this comment (https://github.com/PHPOffice/PhpSpreadsheet/issues/456#issuecomment-389672103), but that also did not help as the solution seems to be out of date with how PHPSpreadsheet works now.

The other issue is that, after attempting to update PHPSpreadsheet with Composer in pursuit of a solution to the above problem, I found that my program no longer functions. It outputs a copy of the original spreadsheet, but opening it returns the error "We found a problem with some content" and trying to recover it shows that the first sheet is wiped totally blank.

Is anyone else seeing any of these  issues lately? Have any solutions be found?

**Comment by oleibman** [comment]
Please upload a copy of the spreadsheet where you see this problem.

**Reply by allredbm** [reply]
[Cable Guide Form.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/7521573/Cable.Guide.Form.xlsx)

Here's a copy of the original excel file that I'm trying to duplicate. The first sheet is where the bulk of the information is being added, and there are formulas that collect that information into strings that will be used for labeling purposes in columns I, P, T, and W. In the other sheets, there are formulas which, when they detect that those columns are filled, fill the cells with that information. In the second issue mentioned in the original post, that first sheet is totally wiped and, while the other sheets are intact, the cells the formulas were watching are blanked out and the formulas are thus rendered useless.

**Reply by oleibman** [reply]
I am sorry, I do not see how this spreadsheet is supposed to work. Can you guide me by telling me what cells to fill with what values, and what cells you expect to see with a new value as a result?

**Reply by allredbm** [reply]
[Cable Guide Form.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/7527874/Cable.Guide.Form.xlsx)

Yeah, I didn't initially write it so I know it's a lot, so here's a version with some test values filled in. In the first sheet, columns E, G, and H being filled cause column I to automatically fill with those values. Same for columns L, N, and O and column P, Columns  R and S for column T, and Columns U and V for column W. Then, on the sheet called 6-54, You can see the values in column B filling with all of those values from the first data row of the first sheet when the requisite cells are filled.

After PHPSpreadsheet writes the copy of the sheet, the formulas still calculate, but only if you click on the cells with the formula and refresh them by pressing Enter. They don't automatically recalculate as they do in the original version that I posted here. I'm trying to get to the bottom of why this is, since in the issue I linked to above it looks like this exact issue was already fixed and a PR was made to add that fix to the codebase. As well, after updating to the most recent version the writing process seems to corrupt that first sheet and wipe it out entirely, so I'm trying to get to the bottom of that as well.

**Reply by oleibman** [reply]
I do not see any cells with "some test values filled in" in your most recent upload.

I took your file, and entered `asset8` in sheet `Cable Request` cell E8. This caused BK8 to be populated with `2`. `asset9` in E9 populated BK9 with `3`. `10` in G8 and `1` in H8 set I8 to `asset8 10.1`. `20` in G9 and `2` in H9 set I9 to `asset9 20.2`. This all seems to be correct according to the formulas in columns BK and I. I did not save the results, so the file I'm using is unchanged.

I then used PhpSpreadsheet to read the unchanged file, and save it to a new file without any changes, I repeated the exercise in the previous paragraph on the new file, and it behaved exactly the same as the original. As soon as I entered the appropriate in columns E, G, and H, columns BK and I showed the new calculated values immediately - no need to click on the cells in BK and I.

Are you seeing different results than I? Or do I need to do something else to demonstrate the problem? If your results are different, perhaps you need to upload the file which you created as a copy. And it would be useful to know PhpSpreadsheet release, PHP release, operating system, and Excel (or equivalent) release.

**Reply by oleibman** [reply]
There is a problem with Data Validations in release 19, which makes the result of loading and saving unreadable with release 19. This does not match your symptom, but, just in case ... There should not be a problem creating the copy with Release 18 or master.

**Reply by allredbm** [reply]
The test values I was referring to were the cells filled in on line 6 of the sheet "Cable Request", the ones that say things like "Test asset" and such. The issues I'm seeing are in the sheet called "6-54". The cells in this sheet are supposed to fill with the values in the first sheet in columns I, P, T, and W respectively when those cells are filled (you can see which cell the row is querying in column S of that sheet. The formula is using those cells to set which column in the "Cable Guide" sheet to watch). By default, this happens automatically and changes to the cells in the "Cable Guide" sheet are reflected in the "6-54" sheet. However, after creating a duplicate using PHPSpreadsheet, I've noticed that the formulas in the "6-54" sheet no longer make these updates automatically and need to be manually prompted by clicking on the cell and hitting Enter to refresh the formula. If you look at sheet 6-54 in your test file, have you noticed anything like this behavior?

**Reply by oleibman** [reply]
When I open the spreadsheets, `6-54` cell B2 contains `Test Host Asset, Test Host Rack, Test Host Port` in both before and after versions. When I change `Cable Request` cell E6 to `Test Host Assets`, that change is reflected in `6-54` B2 in both before and after versions, without any additional actions for either. Likewise, when I fill in E7/G7/H7, those updates are immediately reflected in `6-54` D2. It sounds like you are saying that your experience is different.

**Reply by allredbm** [reply]
Yes, the experience I'm having is different. In my case, after creating the copy, those cells in `6-54` don't refresh automatically. 

Can you elaborate on the process you're using in PHP to create the copy? Maybe I'm doing something wrong. 

```
require '../../vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\Reader\Xlsx as Reader;
use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx as Writer;

$fileLocation = [path to file];

ini_set('max_execution_time', 60);

$reader = new Reader();
$spreadsheet = $reader->load($fileLocation );

$writer = new Writer($spreadsheet);
$fileName = "CABLE DOC COPY.xlsx";
$writer->save($fileName);
```

**Reply by oleibman** [reply]
I tried your sample code. Had to comment out the ini_set statement. Got the expected results. Used PHP7.4, PhpSpreadsheet 19, Windows 10 (OS probably doesn't matter) to generate the output file, Excel 365 Enterprise to examine the results.

**Reply by allredbm** [reply]
[CADDS CABLE DOC - APP3445 (2).xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/7540938/CADDS.CABLE.DOC.-.APP3445.2.xlsx)

This is pretty frustrating. Here's an example of what I'm getting from the same process. If you look at `6-54`, you'll see that B2-B5 isn't even calculating the formula, it just presents the formula as a literal string. The rest all precalculate as expected, but all of the cells do not automatically recalculate the formula when the cells they're looking at change. 

Also, the version of PHPSpreadsheet I'm using is 1.18

**Reply by oleibman** [reply]
I fear this will frustrate you even more. I changed E6 and H6 on the first sheet of the file which you just uploaded; as you can see from the screenshot which I hope I will be able to post correctly, they are immediately reflected in B2 on the second sheet (I did auto-size the row-heights on the second sheet before entering anything on the first).
![image](https://user-images.githubusercontent.com/10341515/141851968-f505001c-7fe4-4630-b20e-62cbc0752937.png)
Perhaps the problem occurs only with a specific version of Excel. As I mentioned, mine is Excel 365 (Microsoft 365 apps for enterprise), Version 2107 Build 14228.20324.

**Reply by allredbm** [reply]
Perhaps it is a version issue. At any rate, it seems  like we've hit bottom as far as investigating this. Thanks for helping me diagnose this anyway.
